src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.cc

```diff
@@ -16,34 +16,329 @@
+  handshaker->interested_parties = interested_parties;
+  handshaker->has_created_handshaker_client = false;
+  // TODO(mattstev): this API is exposed in a PR that is not yet merged.
+  // handshaker->options = grpc_s2a_credentials_options_copy(options);
+  handshaker->base.vtable = &handshaker_vtable;
+
+  *self = &(handshaker->base);
+  return TSI_OK;
+}
+
+static tsi_result s2a_handshaker_result_extract_peer(
+    const tsi_handshaker_result* self, tsi_peer* peer) {
+  // TODO(mattstev): implement.
   return TSI_UNIMPLEMENTED;
 }
 
+static tsi_result s2a_handshaker_result_create_zero_copy_grpc_protector(
+    const tsi_handshaker_result* self, size_t* max_output_protected_frame_size,
+    tsi_zero_copy_grpc_protector** protector) {
+  // TODO(mattstev): the implementation is blocked because the necessary API's
+  // are exposed in a PR that is not yet merged.
+  return TSI_UNIMPLEMENTED;
+}
+
+static tsi_result s2a_handshaker_result_get_unused_bytes(
+    const tsi_handshaker_result* self, const unsigned char** bytes,
+    size_t* bytes_size) {
+  if (self == nullptr || bytes == nullptr || bytes_size == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultUnusedBytesNullptr);
+    return TSI_INVALID_ARGUMENT;
+  }
+  s2a_tsi_handshaker_result* result =
+      reinterpret_cast<s2a_tsi_handshaker_result*>(
+          const_cast<tsi_handshaker_result*>(self));
+  *bytes = result->unused_bytes;
+  *bytes_size = result->unused_bytes_size;
+  return TSI_OK;
+}
+
+static void s2a_handshaker_result_destroy(tsi_handshaker_result* self) {
+  if (self == nullptr) {
+    return;
+  }
+  s2a_tsi_handshaker_result* result =
+      reinterpret_cast<s2a_tsi_handshaker_result*>(self);
+  gpr_free(result->in_traffic_secret);
+  gpr_free(result->out_traffic_secret);
+  gpr_free(result->spiffe_id);
+  gpr_free(result->hostname);
+  gpr_free(result->unused_bytes);
+  gpr_free(result);
+}
+
+static const tsi_handshaker_result_vtable s2a_result_vtable = {
+    s2a_handshaker_result_extract_peer,
+    s2a_handshaker_result_create_zero_copy_grpc_protector, nullptr,
+    s2a_handshaker_result_get_unused_bytes, s2a_handshaker_result_destroy};
+
+// TODO(mattstev): I don't use the application protocol and local identity
+// fields from |response|, or the attributes field from the peer identity.
 tsi_result s2a_tsi_handshaker_result_create(s2a_SessionResp* response,
                                             bool is_client,
-                                            tsi_handshaker_result** result) {
-  return TSI_UNIMPLEMENTED;
+                                            tsi_handshaker_result** self) {
+  if (self == nullptr || response == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultNullptrArguments);
+    return TSI_INVALID_ARGUMENT;
+  }
+  const s2a_SessionResult* handshake_result = s2a_SessionResp_result(response);
+  if (handshake_result == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultEmpty);
```

Comment:
Is there any reason to use a const variable for error message instead of hard-coding it? It's a little inconvenient if the reader wants to know what the error messages are.

Reply:
This was request by Cesar in another PR. (I've tried to make the error message names as explanatory as possible to help with readability.)

Reply:
SG.